url90b6689baa_adsonrocha_PD321_tgz-pJ8-MyFuzzyLiteClassJ8/src/com/fuzzylite/rule/Antecedent.java

```diff
@@ -0,0 +1,271 @@
+import com.fuzzylite.norm.TNorm;
+import com.fuzzylite.term.Function;
+import com.fuzzylite.variable.InputVariable;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.StringTokenizer;
+
+/**
+ *
+ * @author jcrada
+ */
+public class Antecedent {
+
+    protected Expression root;
+
+    public Expression getRoot() {
+        return this.root;
+    }
+
+    public double activationDegree(TNorm conjunction, SNorm disjunction) {
+        return this.activationDegree(conjunction, disjunction, root);
+    }
+
+    public double activationDegree(TNorm conjunction, SNorm disjunction, Expression node) {
+        if (node instanceof Proposition) {
+            Proposition proposition = (Proposition) node;
+            if (!proposition.variable.isEnabled()) {
+                return 0.0;
+            }
+            if (!proposition.getHedges().isEmpty()) {
+                int lastIndex = proposition.getHedges().size() - 1;
+                Hedge any = proposition.getHedges().get(lastIndex);
+                if (any instanceof Any) {
+                    double result = any.hedge(Double.NaN);
+                    while (--lastIndex >= 0) {
+                        result = proposition.getHedges().get(lastIndex).hedge(result);
+                    }
+                    return result;
+                }
+
+            }
+
+            InputVariable inputVariable = (InputVariable) proposition.getVariable();
+            double result = proposition.getTerm().membership(inputVariable.getInputValue());
+            for (int i =  proposition.getHedges().size() - 1 ;i >= 0; --i) {
+                result = proposition.getHedges().get(i).hedge(result);
+            }
+            return result;
+        } else if (node instanceof Operator) {
+            Operator operator = (Operator) node;
+            if (operator.getLeft() == null || operator.getRight() == null) {
+                throw new RuntimeException("[syntax error] left and right operators cannot be null");
+            }
+            if (Rule.FL_AND.equals(operator.getName())) {
+                return conjunction.compute(
+                        activationDegree(conjunction, disjunction, operator.getLeft()),
+                        activationDegree(conjunction, disjunction, operator.getRight()));
+            }
+            if (Rule.FL_OR.equals(operator.getName())) {
+                return disjunction.compute(
+                        activationDegree(conjunction, disjunction, operator.getLeft()),
+                        activationDegree(conjunction, disjunction, operator.getRight()));
+            }
+            throw new RuntimeException(String.format(
+                    "[syntax error] operator <%s> not recognized",
+                    operator.getName()));
+        } else {
+            throw new RuntimeException("[expression error] unknown instance of Expression");
+        }
+    }
+
+    public void load(String antecedent, Engine engine) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 23. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.